web_pwa_oca/templates/service_worker.xml

```diff
@@ -0,0 +1,40 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <template id="pwa_init" name="PWA Initialization">
+        const oca_pwa = new PWA(...<t t-esc="str(pwa_params)"/>);
+    </template>
+    <template id="pwa_core_events" name="PWA Core Events">
+        <t t-set="evt_install">
+            evt.waitUntil(oca_pwa.installWorker());
+            self.skipWaiting();
+        </t>
+        self.addEventListener('install', evt =&gt; {
+            console.log('[ServiceWorker] Installing...');
+            <t t-raw="evt_install" />
+        });
+
+        <t t-set="evt_fetch">
+            evt.respondWith(oca_pwa.processRequest(evt.request));
+        </t>
+        self.addEventListener('fetch', evt =&gt; {
+            console.log('[ServiceWorker] Fetching...');
+            <t t-raw="evt_fetch" />
+        });
+
+        <t t-set="evt_activate">
+            console.log('[ServiceWorker] Activating...');
+            evt.waitUntil(oca_pwa.activateWorker());
+            self.clients.claim();
+        </t>
+        self.addEventListener('activate', evt =&gt; {
+            <t t-raw="evt_activate" />
+        });
+    </template>
+    <template id="service_worker" name="PWA Service Worker">
+        self.importScripts(...<t t-esc="str(pwa_scripts)" />);
```

Comment:
This structure seems very delicate... I see many places where bad code injections can break things... :roll_eyes: 

Besides, we lose all features as linters or autoformatters for JS. And extending qweb text templates is very hard.

Have you considered this?

1. Create a JSON controller that returns all PWA parameters.
1. Have a "normal" JS file that loads that JSON with AJAX and loads the PWA.

Reply:
In this line i use 'Spread Operator': https://developer.mozilla.org/es/docs/Web/JavaScript/Referencia/Operadores/Sintaxis_Spread

Since we don't have 'Odoo Bootstrap' on SericeWorkers i don't see a better way to have a inheritable design. If i put the 'service worker' on a .js file we lost the possibility to extend these service worker initialization process.

Points to know with service worker:
- They have the same scope where are loaded: If you load /blabla/src/js/myserviceworker.js the scope will be '/blablabla/src/js', we want be on main scope ('/').
- The events assignment must be on the main service worker script

Other point is that with these changes we need wait to the json-rpc reply and this introduce a promise on the service worker registration process.

Reply:
Thanks for the details, they help to review! :blush: 

> Other point is that with these changes we need wait to the json-rpc reply and this introduce a promise on the service worker registration process.

I understand.

If I'm not wrong, these processes are executed just once, when the user installs the app, right? :thinking:  If that's the case, then having to wait for some milliseconds until a config JSON file is downloaded from a controller doesn't seem like a big deal, don't you think? The benefits would be still more IMHO.

Reply:
Can be good for some very specific parts... but more of them are incompatible with the service worker because touch DOM parts.

I will add a note in the roadmap.